Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/DeviceStorage.kt

```diff
@@ -0,0 +1,117 @@
+    private val apiLevel: ApiLevel,
+    private val statsFsProvider: StatsFsProvider
+) {
+
+    private val privateStorage = context.filesDir
+    private val storageManager by lazy { context.getSystemService(Context.STORAGE_SERVICE) as StorageManager }
+
+    @TargetApi(Build.VERSION_CODES.O)
+    private fun requestStorageAPI26Plus(targetPath: File, requiredBytes: Long = -1L): CheckResult {
+        Timber.tag(TAG).v(
+            "requestStorageAPI26Plus(path=%s, requiredBytes=%d)", targetPath, requiredBytes
+
+        )
+        val statsManager =
+            context.getSystemService(Context.STORAGE_STATS_SERVICE) as StorageStatsManager
+
+        val storageUUID: UUID = storageManager.getUuidForPath(targetPath)
+
+        val totalBytes = statsManager.getTotalBytes(storageUUID)
+        var availableBytes = statsManager.getFreeBytes(storageUUID)
+
+        if (availableBytes < requiredBytes) {
+            val allocatableBytes = storageManager.getAllocatableBytes(storageUUID)
+            if (allocatableBytes + availableBytes >= requiredBytes) {
+                val toAllocate = requiredBytes - availableBytes
+                Timber.tag(TAG).v(
+                    "Not enough free space, allocating %d on %s.", requiredBytes, targetPath
+                )
+                storageManager.allocateBytes(storageUUID, toAllocate)
+                availableBytes += toAllocate
+            }
+        }
+
+        return CheckResult(
+            path = targetPath,
+            isSpaceAvailable = availableBytes >= requiredBytes || requiredBytes == -1L,
+            freeBytes = availableBytes,
+            totalBytes = totalBytes
+        )
+    }
+
+    private fun requestStorageLegacy(targetPath: File, requiredBytes: Long = -1L): CheckResult {
+        Timber.tag(TAG).v(
+            "requestStorageAPI26Plus(path=%s, requiredBytes=%d)", targetPath, requiredBytes
+        )
+
+        val stats = statsFsProvider.createStats(targetPath)
+
+        return CheckResult(
+            path = targetPath,
+            isSpaceAvailable = stats.availableBytes >= requiredBytes || requiredBytes == -1L,
+            freeBytes = stats.availableBytes,
+            totalBytes = stats.totalBytes
+        )
+    }
+
+    private fun checkSpace(path: File, requiredBytes: Long = -1L): CheckResult =
+        try {
+            val result = if (apiLevel.hasAPILevel(Build.VERSION_CODES.O)) {
+                requestStorageAPI26Plus(path, requiredBytes)
+            } else {
+                requestStorageLegacy(path, requiredBytes)
+            }
+            Timber.tag(TAG).d("Requested %d from %s: %s", requiredBytes, path, result)
+            result
+        } catch (e: Exception) {
+            Timber.tag(TAG).e(e, "Failed to request %d on %s", requiredBytes, path)
+            // Android storage is a fickle beast
+            // If the space request fails, it's possible that the write attempt will still succeed
+            CheckResult(path, true, -1L, -1L)
```

Comment:
What if the write attempt fails, how is the use planned?

Reply:
My concerns were that it has a slightly higher chance of failing via `StorageManager.getUuidForPath` in cases where the OEM did some funky things with the sdcard and the user swaps internal/external storage. I've already seen cases where they didn't give storage valid IDs.

You are right, this could be improved, instead of swallowing the error, we should fall back to the <API26 method, which has a very low chance of failing, and if that also fails, throw an exception (documented) and then the caller has to determine how to handle it.

Reply:
https://github.com/corona-warn-app/cwa-app-android/pull/1076/commits/bd244907317c1b98f4fd0685fd51667391fb283d